### ISSUE TITLE
KeePassXC Browser doesn't detect new/changed credentials!

### ISSUE THREAD

**Issue by DocMAX** [keepassxreboot/keepassxc-browser#67]
## Expected Behavior
Detect credentials change

## Current Behavior
Ignores changes

## Possible Solution
Detect new/changed credentials

## General Info
 You are running KeePassXC version: 2.3.1
Latest available version of KeePassXC: 2.3.1 

Operating system: WIN7
Browser: FIREFOX 58
Proxy used: NO

**Comment by varjolintu** [comment]
In which site this happens? Can you give some info do you change the username, password or both when this occurs?

**Comment by robert-franz** [comment]
I also can verify this issue. Working with Firefox 59.0.2, Debian 8, KeePassXC 2.3.1, KeePassXC-Browser 1.0.1. It's not possible to change the password as the change is not stored in the database.

**Comment by varjolintu** [comment]
Please check the following:
- Database you want to save credentials is opened
- Database open is connected to the browser extension
- You can receive credentials and autocomplete or filling them from context menu works

**Comment by DocMAX** [comment]
We checked this allready. I'm back at original KeePass/HTTP for now.

**Comment by varjolintu** [comment]
No we didn't. I cannot reproduce this.

**Comment by robert-franz** [comment]
@varjolintu All the three steps work properly. Simply storing changed credentials into the database via keepassxc-browser does not work. In case of changing the credentials in the database via keepassxc, those changed credentials are available directly via keepassxc-browser.

When entering different credentials this popup appears:
![keepassxc](https://user-images.githubusercontent.com/13948222/38495728-efca3266-3bfa-11e8-8529-d2b6233dd93f.png)

But when clicking the activate button or the extension icon there is no way to store the changed password. Also  the password is not stored automatically.

![keepassxc1](https://user-images.githubusercontent.com/13948222/38495838-6c743bea-3bfb-11e8-9b4d-07536ecc2d6a.png)

**Comment by varjolintu** [comment]
@robert-franz Does the extension icon blink for several seconds? You have to click it while this happens. If the notification is annoying or it covers the icon, you can disable notifications from the settings.

**Comment by Frederick888** [comment]
Exactly the same issue here. Firefox 60.0b10 with KDE Plasma under Linux.

**Comment by Frederick888** [comment]
Hmm... I found that actually I need to click the icon for a second time to make the new/update credentials page show up.

I turned on logging for the event loop...
```diff
 kpxcEvent.onMessage = function(request, sender, callback) {
     if (request.action in kpxcEvent.messageHandlers) {
-        //console.log('onMessage(' + request.action + ') for #' + sender.tab.id);
+        if (request.action != 'check_databasehash' && request.action != 'update_available_keepassxc') {
+            console.log('onMessage(' + request.action + ')');
+        }
```
...and got the following outputs
```
// click login page link
onMessage(load_settings)
onMessage(page_clear_logins)
onMessage(retrieve_credentials)
onMessage(popup_login)
onMessage(page_clear_logins)
Error: Promised response from onMessage listener went out of scope

// click submit button
onMessage(set_remember_credentials)
JQMIGRATE: Migrate is installed, version 1.4.1
onMessage(load_settings)
onMessage(page_clear_logins)
onMessage(page_clear_logins)

// first click on icon
onMessage(get_status)

// second click on icon
onMessage(stack_add)
onMessage(get_tab_information)
onMessage(get_connected_database)
```
Why do those event only arrive after the second click? And @robert-franz , does a second click work for you?

The icon page updates in a somewhat weird way... After `popup_remember.html` has shown up, the first click on "Update" does all the jobs normally, but I need a second click on it to get back to the default view (the one with "Redetect ..." button).

**Comment by varjolintu** [comment]
The second click happens on `popup_remember.html` and the first one in `popup.html`. Still, at the submit the `popup_remember.html` is already set by `onMessage(set_remember_credentials)`. You could check what happens in `browserAction.setRememberPopup` after submit.

For me, this issue is still very hard to reproduce. It's probably related to the dynamic switching between different popups.

**Comment by robert-franz** [comment]
I tried the things you described above. When double click on the blinking icon directly after entering the new credentials, the new credentials are stored within the keepassxc database. So this works. But imho this behaviour is really confusing ...

**Comment by varjolintu** [comment]
Confusing for me too. I'm not sure why I cannot reproduce this behaviour. I have tried to do it with all three major OS's. I also recommend that you increase the Redirect Allowance value from the settings and see if it affects to anything.

**Comment by Frederick888** [comment]
Ok, good news. I found that if I add a breakpoint at `browserActions.js:240`
```javascript
        browserAction.show(null, {'id': id});
```
...and hold it until the new page finish loading, I then don't need a second click to see the `popup_remember.html` view. Solution is simple: move it to `browser.tabs.onUpdated`.

The second click for the "Update" button is actually a different issue &mdash; `_close();` is missing after the click.

So here's the patch. @robert-franz if you could check whether it works for you :)
@varjolintu I'm a bit unsure about the multi-credential part for the update button, does it need a `_close()` as well somewhere? Please review the patch, thanks.

```diff
diff --git a/keepassxc-browser/background/browserAction.js b/keepassxc-browser/background/browserAction.js
index d2fdb7d..35e5109 100755
--- a/keepassxc-browser/background/browserAction.js
+++ b/keepassxc-browser/background/browserAction.js
@@ -236,10 +236,8 @@ browserAction.setRememberPopup = function(tabId, username, password, url, userna
             usernameExists: usernameExists,
             list: credentialsList
         };
 
-        browserAction.show(null, {'id': id});
-
         if (page.settings.showLoginNotifications) {
             showNotification('Create or modify the credentials by clicking on the extension icon.');
         }
     });
diff --git a/keepassxc-browser/background/init.js b/keepassxc-browser/background/init.js
index 4f6c6af..5c14740 100644
--- a/keepassxc-browser/background/init.js
+++ b/keepassxc-browser/background/init.js
@@ -69,8 +69,9 @@ browser.tabs.onActivated.addListener((activeInfo) => {
  * @param {object} changeInfo
  */
 browser.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
     if (changeInfo.status === 'complete') {
+        browserAction.show(null, {'id': tabId});
         kpxcEvent.invoke(browserAction.removeRememberPopup, null, tabId, []);
     }
 });
 
diff --git a/keepassxc-browser/popups/popup_remember.js b/keepassxc-browser/popups/popup_remember.js
index be79328..80b2e38 100644
--- a/keepassxc-browser/popups/popup_remember.js
+++ b/keepassxc-browser/popups/popup_remember.js
@@ -23,8 +23,9 @@ function _initialize(tab) {
         browser.runtime.sendMessage({
             action: 'add_credentials',
             args: [_tab.credentials.username, _tab.credentials.password, _tab.credentials.url]
         }).then(_verifyResult);
+        _close();
     });
 
     $('#btn-update').click(function(e) {
         e.preventDefault();
@@ -34,8 +35,9 @@ function _initialize(tab) {
             browser.runtime.sendMessage({
                 action: 'update_credentials',
                 args: [_tab.credentials.list[0].uuid, _tab.credentials.username, _tab.credentials.password, _tab.credentials.url]
             }).then(_verifyResult);
+            _close();
         }
         else {
             $('.credentials:first .username-new:first strong:first').text(_tab.credentials.username);
             $('.credentials:first .username-exists:first strong:first').text(_tab.credentials.username);
```

**Comment by droidmonkey** [comment]
Please submit as a pull request. Great work!

**Comment by varjolintu** [comment]
@Frederick888 Thanks for the info. In `popup_remember.js` the `_verifyResult` already calls `_close()` if everything works fine. Are you sure you need those extra calls? Because those are asynchronous calls, `_close()` must be wrapped inside `catch()`.

Also, I don't think moving the `browserAction.show()` is a correct solution for this. But now I have a better understanding why this may happen. So thank you for that.

**Comment by Frederick888** [comment]
@varjolintu Do you still need a PR or I can now lie back and wait for it to be fixed? 😆

**Comment by varjolintu** [comment]
@Frederick888 Just lie back :) The thing I noticed is that at error situations `_verifyResult()` is not called at all, and `catch()` also doesn't catch the error. So I'm checking the function handling at `keepass.js` if something's missing. It would be also nice to show a notification when saving or updating credentials fail.

And btw, is your problem fixed if you only comment out the `browserAction.show(null, {'id': id});` line and do nothing else?

**Comment by Frederick888** [comment]
@varjolintu Just about to say that I couldn't get the breakpoint in `_verifyResult` after the first "Update" click lol. It seems that you have come up with a far less hacky solution, thanks a lot.

And no, that line can be kept and it still works... simply would like to avoid duplication.

**Comment by varjolintu** [comment]
Here's my suggestion for the fix, please try it out:
```diff
diff --git a/keepassxc-browser/background/browserAction.js b/keepassxc-browser/background/browserAction.js
index d2fdb7d..35e5109 100755
--- a/keepassxc-browser/background/browserAction.js
+++ b/keepassxc-browser/background/browserAction.js
@@ -237,8 +237,6 @@ browserAction.setRememberPopup = function(tabId, username, password, url, userna
             list: credentialsList
         };

-        browserAction.show(null, {'id': id});
-
         if (page.settings.showLoginNotifications) {
             showNotification('Create or modify the credentials by clicking on the extension icon.');
         }
```

```diff
diff --git a/keepassxc-browser/popups/popup_remember.js b/keepassxc-browser/popups/popup_remember.js
index be79328..c53de7c 100644
--- a/keepassxc-browser/popups/popup_remember.js
+++ b/keepassxc-browser/popups/popup_remember.js
@@ -91,9 +91,10 @@ function _connected_database(db) {
 }

 function _verifyResult(code) {
-    if (code === 'success') {
-        _close();
+    if (code === 'error') {
+        showNotification('Error: Credentials cannot be saved or updated.');
     }
+    _close();
 }
```

**Comment by varjolintu** [comment]
@Frederick888 Any error messages in the console after the first click? You can also try to set a breakpoint to the `keepass.js:173` where the response should arrive.

**Comment by Frederick888** [comment]
@varjolintu 
1. No, I got no error message through the whole procedure.
0. I set a breakpoint at `keepass.js:173` and it normally goes the way down to `callback(keepass.verifyResponse(parsed, incrementedNonce) ? 'success' : 'error');` (confirmed the argument is `'success'`), but then my another breakpoint in `_verifyResult` just doesn't get triggered. It only works after a second click.

**Comment by varjolintu** [comment]
I really wonder why that happens. There's no reason. Does this happen with both Chromium (or Chromium-based browser) and Firefox?

**Comment by Frederick888** [comment]
@varjolintu Errrr.... I have only Google Chrome and Firefox Developer Edition installed so I tried it in Chrome. But after loading it in through `Extensions -> LOAD UNPACKED` it says that it cannot connect to KeePassXC. The one from Chrome web store works normally tho. I'm using VS Code, is it possible to start a debug session directly from the IDE?

And I also tested the `.catch()`... but no, still no error messages.

**Comment by varjolintu** [comment]
@Frederick888 You need to add the new extension ID (it's generated if you load it manually) to the native messaging json script. I've never tried anything browser related with VS Code so cannot help you there.

**Comment by Frederick888** [comment]
Got the same results with the patches in Google Chrome 65.0.3325.181 (Official Build) (64-bit) under Linux. @varjolintu

**Comment by Frederick888** [comment]
Maybe I've done something stupid???? Could you pack the extension for either Firefox or Chrome and attach it here?

**Comment by varjolintu** [comment]
@Frederick888 I've tried against many websites. I also tried against wordpress.com and an instance that uses WordPress. Are there any other pages this error happens? If not, you can make an account for me using my email address sami.vanttinen@protonmail.com and mail me the URL so it's kept private.

**Comment by varjolintu** [comment]
@Frederick888 Tried it but still cannot reproduce it. Do you mean a situation where you change your password (or it's not saved) and you want to save it after that from the popup?

**Comment by Frederick888** [comment]
@varjolintu Yup
Steps:
1. Sign in to save the credentials to KP first
2. Change the password, do not touch KP icon, log out, close the tab straightaway
3. Go to the website again, use new password to log in
4. Logged in, page loaded, click browser KP icon, there come the problems

**Comment by varjolintu** [comment]
Finally got it reproduced.

**Comment by varjolintu** [comment]
Here are the changes I made to get this work:
https://github.com/keepassxreboot/keepassxc-browser/compare/remember_popup_fix?expand=1

You were right about the correct location of `browserAction.show()`. Please test this.

**Comment by varjolintu** [comment]
And btw, that patch is not ready. Moving the `browserAction.show()` causes the popup state to change. For example, if you haven't got any credentials saved for the site, the icon stays gray with a cross. And if you already had credentials but you're now updating it, the popup stays at `popup_login` and shows an empty list. So there's a lot to do.

**Comment by Frederick888** [comment]
Good to know you've addresses the issue! It's definitely better to get it fixed from the root rather than an ugly hack. I'll stay in loop and if you need any help just shout. Thanks for your great work!

**Comment by varjolintu** [comment]
I also found some issues with the options saving. Some integers were saved as strings but still used as integers later etc.. messy stuff. While JavaScript "allows" this, it's still wrong. So that will be fixed too. Thanks for helping solving this one!

**Comment by Frederick888** [comment]
@varjolintu I'm afraid it still persists. Using KeePassXC 2.3.3 and KeePassXC-Browser 1.1.3 with Firefox 61.0b3.

**Comment by varjolintu** [comment]
@Frederick888 Thank you for testing it!

**Comment by getkey** [comment]
It solved the problem for me! Thanks @varjolintu!
I had the issue on Linux only (I'm no longer at my computer so I can't tell you my version numbers), not on macOS.

**Comment by Frederick888** [comment]
Well, I just tested in macOS with the same setup and reproduced the issue.

**Comment by varjolintu** [comment]
@Frederick888 Can you try [this branch](https://github.com/keepassxreboot/keepassxc-browser/tree/remember_popup_fix) manually if it finally fixes the issue? It seems to work for me.

**Comment by Frederick888** [comment]
@varjolintu Yes it does! 👏

**Comment by justintime4tea** [comment]
If anyone comes here looking for a potential fix, and the links within this issue do not resolve, here is one more thing to try:

#### TLDR;
* Make sure a group exists with the name `KeePassXC-Browser Passwords`, within the connected KeePassXC database
  * If this group does not exist, create it.
  * After this group exists, re-login or complete a new registration to trigger KeePassXC update/save prompt.

#### Further details
|:--|---:|
|:--|---:|
|Symptom|User log-in save (new, from KeePassXC prompt) seemingly does not work; no user is created inside KeePassXC although all other features and integrations work.|
|Distribution|Linux fedora 5.19.8-200.fc36.x86_64 |
|Firefox version|firefox-104.0.2-1.fc36.x86_64|
|KeepassXC version|keepassxc-2.7.1-11.fc36.x86_64|
|Method of launch|~/.local/share/applications/firefox.desktop|
|Notes|All other features and integrations seemingly worked, except new user creation|

#### Solution
* Confirm there is no group called `KeePassXC-Browser Passwords`
  * This is the default group (at time of writing) which is used by Firefox when creating new entries.
  * Either this folder was never created or I deleted it and just don't recall doing so.
* **Create a new group and name it as shown above; `KeePassXC-Browser Passwords`**
  * I was able to create this group in the main/root group/section of KeePassXC, if you try a sub-folder, and it doesn't work, I would suggest create the group in main/root.
* Go back to the website that you wanted to create credentials for and either register, finish register, or log in and follow the KeePassXC save
  * If you had finished registration and are currently logged in, log out and then log back in to trigger KeePassXC update/save prompt.
  * This entry should now exist within.

**Comment by varjolintu** [comment]
@justintime4tea KeePassXC should create the default `KeePassXC-Browser Passwords` group if it doesn't exist and user is trying to save to it (using extension default settings).

**Comment by justintime4tea** [comment]
@varjolintu that's what I had assumed would happen if it needed that password group to exist. As I mention in my breakdown, I very well may have deleted the group myself, but even so, it didn't get generated for me automatically when attempting to save a new set of creds. Perhaps the creation of this folder only occurs once during some initialize/first-setup process and never again after? I could see how people not realizing what it is for "cleaning" up and either renaming or deleting the default password group. Either way, thanks for the reply and feedback on intended behavior!

**Comment by varjolintu** [comment]
@justintime4tea Gotta check if there's some certain pattern where the group is not created. If so, it's clearly a bug.

**Comment by justintime4tea** [comment]
I will try a stab at a repro. I may be able to even pull the source and try and step through some of the flows myself. I would love to contribute but time is something I have very little of unfortunately :(

Thank you again for your responsiveness and care on the matter.
